lib/VCMIDirs.cpp

```diff
@@ -164,178 +97,111 @@ class VCMIDirsWIN32 final : public IVCMIDirs
 		std::string libraryName(const std::string& basename) const override;
 
 		void init() override;
+
 	protected:
-		bfs::path oldUserDataPath() const;
-		bfs::path oldUserSavePath() const;
+		mutable std::optional<JsonNode> dirsConfig;
+		void loadDirsJsonIfNeeded() const;
+		bfs::path getPathFromConfigOrDefault(const std::string& key, const std::function<bfs::path()>& fallbackFunc) const;
+		std::wstring utf8ToWstring(const std::string& str) const;
 };
 
-void VCMIDirsWIN32::init()
+std::wstring VCMIDirsWIN32::utf8ToWstring(const std::string& str) const
 {
-	// Call base (init dirs)
-	IVCMIDirs::init();
-
-	// Moves one directory (from) contents to another directory (to)
-	// Shows user the "moving file dialog" and ask to resolve conflits.
-	// If necessary updates current directory.
-	auto moveDirIfExists = [](const bfs::path& from, const bfs::path& to) -> bool
+	std::wstring result;
+	int size_needed = MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, nullptr, 0);
+	if (size_needed > 0)
 	{
-		if (!bfs::is_directory(from))
-			return true; // Nothing to do here. Flies away.
-
-		if (bfs::is_empty(from))
-		{
-			if (bfs::current_path() == from)
-				bfs::current_path(to);
+		result.resize(size_needed - 1);
+		MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, &result[0], size_needed);
+	}
+	return result;
+}
 
-			bfs::remove(from);
-			return true; // Nothing to do here. Flies away.
-		}
+void VCMIDirsWIN32::loadDirsJsonIfNeeded() const
+{
+	if (dirsConfig.has_value())
+		return;
```

Comment:
imagine user creates this file in the middle of the game session. I believe something may go terribly wrong in such case.

Reply:
I think everything is already loaded during game session?

Reply:
I haven't checked deeply code paths when this new method is called. When VCMI requests one of the paths, it shouldn't call this method, otherwise you may get different results during runtime.

Reply:
Loading dirs.json is now moved to constructor, so this behavior should not happend.